### ISSUE TITLE
Automatic Transitive Deployment for NPM & Bower WebJars

### ISSUE THREAD

**Issue by blast-hardcheese** [webjars/webjars#1186]
After successfully packaging `"org.webjars.npm" % "multiline" % "1.0.2"` and including it in my `build.sbt`, I get:

```
[vocities] $ run 9001
[info] Updating {file:/home/blast/Projects/vocities/vocities/}root...
[info] Resolving org.webjars.npm#strip-indent;[1.0.0,2) ...
[warn]  module not found: org.webjars.npm#strip-indent;[1.0.0,2)
[warn] ==== local: tried
[warn]   /home/blast/.ivy2/local/org.webjars.npm/strip-indent/[revision]/ivys/ivy.xml
[warn] ==== activator-launcher-local: tried
[warn]   file:/home/blast/.tools/localized/activator-1.3.2-minimal/repository/org.webjars.npm/strip-indent/[revision]/ivys/ivy.xml
[warn] ==== public: tried
[warn]   http://repo1.maven.org/maven2/org/webjars/npm/strip-indent/[revision]/strip-indent-[revision].pom
[warn] ==== typesafe-releases: tried
[warn]   https://repo.typesafe.com/typesafe/releases/org/webjars/npm/strip-indent/[revision]/strip-indent-[revision].pom
[warn] ==== typesafe-ivy-releasez: tried
[warn]   https://repo.typesafe.com/typesafe/ivy-releases/org.webjars.npm/strip-indent/[revision]/ivys/ivy.xml
[warn] ==== Typesafe Releases Repository: tried
[warn]   https://repo.typesafe.com/typesafe/releases/org/webjars/npm/strip-indent/[revision]/strip-indent-[revision].pom
[warn] ==== sonatype-releases: tried
[warn]   https://oss.sonatype.org/content/repositories/releases/org/webjars/npm/strip-indent/[revision]/strip-indent-[revision].pom
[warn] ==== google-sedis-fix: tried
[warn]   http://pk11-scratch.googlecode.com/svn/trunk/org/webjars/npm/strip-indent/[revision]/strip-indent-[revision].pom
[info] Resolving jline#jline;2.11 ...
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  ::          UNRESOLVED DEPENDENCIES         ::
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[warn]  :: org.webjars.npm#strip-indent;[1.0.0,2): not found
[warn]  ::::::::::::::::::::::::::::::::::::::::::::::
[trace] Stack trace suppressed: run last *:update for the full output.
[error] (*:update) sbt.ResolveException: unresolved dependency: org.webjars.npm#strip-indent;[1.0.0,2): not found
[error] Total time: 8 s, completed Aug 29, 2015 3:27:18 PM
```

I'm now manually packaging `strip-indent@1.0.0`, but it should have been automatically picked up and packaged, imho.

**Comment by jamesward** [comment]
Thanks for filing an issue for this.  I've wanted to do automatic transitive deployment but it will take some work to deal with cyclical references in the graph.

**Comment by blast-hardcheese** [comment]
Happy to contribute! Thanks for providing this service!

Is there an interim solution before the cyclical references are resolved? Maybe a note at the end of deployment suggesting other things that might need to be deployed based purely on which deps 404 when trying to get package info from the repo?

**Comment by jamesward** [comment]
We could add a message stating that transitive dependencies haven't been automatically deployed.  But figuring out which transitives haven't been deployed is the same problem as deploying them.

**Comment by blast-hardcheese** [comment]
I was just thinking checking immediate dependencies, since that's a finite and well defined list

**Comment by schmitch** [comment]
actually does this issue mean that if I add webpack that all transitive dependencies will be stored in org.webjars.npm % webjars or does that mean that all transitive npm packages will be deployed as their own webjar?

Currently It would've been way better to follow npm so that the webjars of webpack has a node_modules folder.

**Comment by jamesward** [comment]
@schmitch This issue is just addressing the web deployment tooling for NPM and Bower WebJars.  So nothing to do with Webpack.

**Comment by Khorkhe** [comment]
@jamesward The same behavior occurs for any NPM module I try to import through WebJars.  I found your blog post about NPM WebJars (published May 2015) and assumed this would work.  Is there something I am missing?
Thanks

**Comment by jamesward** [comment]
@Khorkhe Deployment is manual for all libraries and their dependencies.  But this is something that anyone can do with NPM & Bower WebJars.

I think a good / easy first step with this is to present a list of missing transitive dependencies when a deployment finishes.  That would make it easy for someone to walk through the tree to manually deploy everything.

**Comment by Khorkhe** [comment]
@jamesward When you say "deployment is manual", do you mean I have to npm init and npm install the packages to use?  If that's the case, I have tried that and here is what I reached:
- using browserify, I was able to "require" npm plugins' javascripts (e.g. jquery, bootstrap ...)
- for CSS however, I tried using css "import" on directory node_modules/... (using npm-css module) as well as standard insertion of external stylesheets.  
Attempting this on a non-playframework project worked (simple index.html with css files).  But once used within my Play app, it would return a 404.  I made sure the path specified included the appropriate amount of "../"

I found an interesting paragraph in Play Migration [docs](https://www.playframework.com/documentation/2.5.x/Migration23)

> npm can be used as well as WebJars by declaring a package.json file in the root of your project. Assets  from npm packages are extracted into the same lib folder as WebJars so that, from a code perspective,  there is no concern whether the asset is sourced from a WebJar or from an npm package.

After browsing project files, I found npm packages were extracted in /node_modules.  Attempting to use WebJar syntax to locate CSS files belonging to an npm module returns a WebJar not found error in the Play app.

Do you happen to have any ideas?
Thanks again.

**Comment by jamesward** [comment]
@Khorkhe WebJars are JS/CSS libs repackaged into Jars and deployed on Maven Central.  That packaging and deployment is manual but once deployed anyone can consume the lib from a JVM package manager (like sbt).  NPM WebJars do this for libs deployed on the NPM repository.

So if you are using `package.json` (which it sounds like you are) then you are using NPM to manage those packages, not WebJars.

Does that help?

**Comment by jamesward** [comment]
@Khorkhe BTW: Probably a better place to discuss this is: https://gitter.im/webjars/webjars

**Comment by Khorkhe** [comment]
@jamesward Ok my last comment here:
I am only using package.json because I needed a package only available as npm webjars (Browserify) and i was failing to import it.

> I think a good / easy first step with this is to present a list of missing transitive dependencies when a deployment finishes. That would make it easy for someone to walk through the tree to manually deploy everything.

I am sorry i'm having a hard time understanding what you mean in that paragraph above:

- after some tests, I find that more common npm webjars (such as jquery, bootstrap and angular) are working out of the box with the provided link on webjars.org (e.g.: "org.webjars.npm" % "jquery" % "2.2.3"), but less common webjars such as Browserify (only available through npm webjars) was causing error.  Does that mean not every npm package listed on webjars.org is readily available for use? In the case of bootstrap, jquery and angular above, someone already packaged/deployed once and made it available for everyone else?
- are there any resources / tutorials / explanations on how to package and deploy an npm package as webjar, so I can do it for Browserify?
Thanks and sorry for the confusion.

**Comment by jamesward** [comment]
It is possible that the WebJar you are trying to use is failing because of missing transitive dependencies.  This usually happens when libraries are missing required metadata so they can't be imported into Maven Central.  It is an unfortunate situation but there isn't much we can do to fix it.

The alternative is to use Classic WebJars instead.  Info on creating Classic WebJars: http://www.webjars.org/contributing

**Comment by mkotsbak** [comment]
@Khorkhe what does the import system say? I usually see that error when the license is not found. In that case, please report the problem to the upstream package or add a PR to the project.

**Comment by jamesward** [comment]
I'm making some progress on this: https://github.com/webjars/webjars/tree/wip-autotrans

**Comment by jamesward** [comment]
This is now deployed on the test server.  Try out deployment via https://webjars-test.herokuapp.com or `curl -X POST 'https://webjars-test.herokuapp.com/deploy?webJarType=bower&nameOrUrlish=bootstrap&version=3.3.1'

Full changes: https://github.com/webjars/webjars/compare/wip-autotrans

**Comment by SomeoneToIgnore** [comment]
Hello, @jamesward 

I've tried to check it with the latest [vaadin-core](https://github.com/vaadin/vaadin-core/releases), but got this:
```
curl -X POST 'https://webjars-test.herokuapp.com/deploy?webJarType=BowerGitHub&nameOrUrlish=vaadin-core&version=10.0.0-alpha15'
Got package info for org.webjars.bowergithub.vaadin vaadin-core 10.0.0-alpha15
Determining dependency graph
curl: (18) transfer closed with outstanding read data remaining
```

And I see no `vaadin-core` here: https://dl.bintray.com/webjars/maven/org/webjars/bowergithub/vaadin/

Am I doing something wrong?

**Comment by jamesward** [comment]
Thanks for giving it a try!  I've fixed a bunch of bugs so this is ready for more testing.

**Comment by SomeoneToIgnore** [comment]
Nicely done, now works, thanks!

A few suggestions:

* Now it takes quite some time to deploy everything as if it was sequential. If that's true, then you can surely improve that one :)
* You still have not got rid of that `v` prefix fully, that's sort of irritating: I have to specify one when using GitHub url when accessing the API (the one I invoke with `curl`) and in the package names, you have a mixture of both:
```
  Deploying these dependencies: vaadin/vaadin-lumo-styles#v1.0.0-alpha3 vaadin/vaadin-grid#v5.0.0-alpha7 vaadin/vaadin-progress-bar#v1.0.0-alpha12 vaadin/vaadin-date-picker#v3.0.0-alpha6 vaadin/vaadin-checkbox#v2.0.0-alpha7 vaadin-themable-mixin#1.1.5 iron-a11y-keys-behavior#2.0.1 iron-iconset-svg#2.1.1 vaadin/vaadin-notification#v1.0.0-alpha7 iron-icon#2.0.1 vaadin/vaadin-themable-mixin#v1.1.5 vaadin/vaadin-upload#v4.0.0-alpha5 webcomponents/webcomponentsjs#v1.1.0 vaadin/vaadin-radio-button#v1.0.0-alpha12 PolymerElements/iron-fit-behavior#v2.1.0 vaadin/vaadin-button#v2.0.0-alpha6 PolymerElements/iron-scroll-target-behavior#v2.0.0 iron-a11y-announcer#2.0.0 polymer#2.5.0 vaadin/vaadin-split-layout#v4.0.0-alpha5 vaadin/vaadin-text-field#v2.0.0-alpha10 vaadin/vaadin-element-mixin#v1.0.2 vaadin/vaadin-item#v2.0.0-alpha5 vaadin/vaadin-tabs#v2.0.0-alpha7 Polymer/polymer#v2.5.0 vaadin/vaadin-dialog#v2.0.0-alpha5 PolymerElements/iron-meta#v2.0.5 vaadin/vaadin-list-box#v1.0.0-alpha9 vaadin/vaadin-combo-box#v4.0.0-alpha6 vaadin/vaadin-icons#v4.1.3 vaadin/vaadin-development-mode-detector#v1.0.3 vaadin/vaadin-form-layout#v2.0.0-alpha6 iron-resizable-behavior#2.0.1 polymerelements/iron-iconset-svg#v2.1.1 vaadin/vaadin-dropdown-menu#v1.0.0-alpha10 vaadin/vaadin-usage-statistics#v1.0.6 iron-list#2.0.12 vaadin/vaadin-ordered-layout#v1.0.0-alpha8 polymerelements/iron-icon#v2.0.1 PolymerElements/iron-a11y-keys-behavior#v2.0.1 iron-media-query#2.0.0 PolymerElements/iron-flex-layout#v2.0.1 vaadin-overlay#2.0.5 webcomponents/shadycss#v1.1.1 vaadin/vaadin-context-menu#v4.0.0-alpha6 PolymerElements/iron-overlay-behavior#v2.3.1 vaadin-control-state-mixin#2.0.3 PolymerElements/iron-resizable-behavior#v2.0.1 vaadin/vaadin-control-state-mixin#v2.0.3
```
* Also note that you have duplicates (have a look at the output from the previous bullet):
`Polymer/polymer#v2.5.0` and `polymer#2.5.0` (note `v` prefixes again)

**Comment by jamesward** [comment]
@SomeoneToIgnore Thanks!  Some feedback:
1. Yes, the dependency deploys are now sequential because there was an issue doing parallel Maven Central deploys.  I've filed a support request with BinTray so hopefully they can resolve that and then we can make it parallel.
1. In the case of GitHub references the specified version must be a tag.  But if you specify a Bower artifact name then the version can be anything that resolves in the Bower repo so either with a `v` or without.  I'm not sure there is a better way to handle this.
1. The dependency list is pre-resolved so `Polymer/polymer#v2.5.0` and `polymer#2.5.0` will resolve to the same artifact but I'll have to make some changes to resolve those when the dependencies are calculated.
1. I'm investigating the exceptions.
1. Maven Central deployment is enabled on the test server right now so we can test the whole thing.
